File: backend/agent/nodes/placeholder_detect.py
import re
from typing import Dict, Any, List

PLACEHOLDER_PATTERNS = [
    r'\'\[.*?\]\'',         # '[last_name]', '[value]'
    r'\'<.*?>\'',           # '<last_name>'
    r'\'placeholder\'',     # literal word placeholder
    r'\'unknown\'',         # literal word unknown
    r'\'your_.*?\'',        # 'your_last_name'
    r'= \'\'',               # empty string filter
    r'\'NULL\'',            # string 'NULL' as value
    r'= \$\w+',            # unreplaced vars like $lastName
    r'\[MISSING.*?\]',    # [MISSING_VALUE]
    r'\[.*?_HERE\]',      # [VALUE_HERE]
]

def detect_placeholders(sql: str) -> List[str]:
    found = []
    for pattern in PLACEHOLDER_PATTERNS:
        matches = re.findall(pattern, sql, re.IGNORECASE)
        if matches:
            found.extend(matches)
    # Remove duplicates preserving order
    seen = set()
    return [x for x in found if not (x in seen or seen.add(x))]

def extract_missing_field_names(placeholders: List[str]) -> List[str]:
    cleaned = []
    for p in placeholders:
        # Strip quote/brackets characters
        f = re.sub(r'[\'\[\]<>]', '', p).strip()
        cleaned.append(f)
    return cleaned

def build_clarification_question(missing_fields: List[str]) -> str:
    if len(missing_fields) == 1:
        return f"To complete this query safely, I need one more detail: what is the {missing_fields[0]}?"
    field_list = ", ".join(missing_fields)
    return f"To complete this query safely, I need a few more details: {field_list}. Could you provide these?"

def placeholder_detect_node(state: Dict[str, Any]) -> Dict[str, Any]:
    generated_sql = state.get("generated_sql")
    if not generated_sql:
        return {}

    placeholders = detect_placeholders(generated_sql)
    if not placeholders:
        return {}  # clean — proceed

    missing_fields = extract_missing_field_names(placeholders)
    question = build_clarification_question(missing_fields)

    return {
        "pending_clarification": {
            "reason": "placeholder",
            "missingFields": missing_fields,
            "partialSQL": generated_sql,  # save for resume
            "question": question
        },
        "final_response": {
            "type": "clarification",
            "message": question
        }
    }

File: lib/agent/nodes/placeholder_detect.ts
// lib/agent/nodes/placeholder_detect.ts
import type { AgentStateType } from "../state";

const PLACEHOLDER_PATTERNS = [
  /'\[.*?\]'/gi, // '[last_name]', '[value]'
  /'<.*?>'/gi, // '<last_name>'
  /'placeholder'/gi, // literal word placeholder
  /'unknown'/gi, // literal word unknown
  /'your_.*?'/gi, // 'your_last_name'
  /= ''/g, // empty string filter
  /'NULL'/gi, // string 'NULL' as value
  /= \$\w+/g, // unreplaced vars like $lastName
  /\[MISSING.*?\]/gi, // [MISSING_VALUE]
  /\[.*?_HERE\]/gi, // [VALUE_HERE]
];

function detectPlaceholders(sql: string): string[] {
  const found: string[] = [];
  for (const pattern of PLACEHOLDER_PATTERNS) {
    const matches = sql.match(pattern);
    if (matches) found.push(...matches);
  }
  return [...new Set(found)];
}

function extractMissingFieldNames(placeholders: string[]): string[] {
  return placeholders.map((p) => p.replace(/['\[\]<>]/g, "").trim());
}

function buildClarificationQuestion(
  missingFields: string[],
  userPrompt: string,
): string {
  if (missingFields.length === 1) {
    return `To complete this query safely, I need one more detail: what is the ${missingFields[0]}?`;
  }
  const fieldList = missingFields.join(", ");
  return `To complete this query safely, I need a few more details: ${fieldList}. Could you provide these?`;
}

export async function placeholderDetectNode(
  state: AgentStateType,
): Promise<Partial<AgentStateType>> {
  if (!state.generatedSQL) return {};

  const placeholders = detectPlaceholders(state.generatedSQL);

  if (placeholders.length === 0) return {}; // clean — proceed

  const missingFields = extractMissingFieldNames(placeholders);
  const question = buildClarificationQuestion(missingFields, state.userPrompt);

  return {
    pendingClarification: {
      reason: "placeholder",
      missingFields,
      partialSQL: state.generatedSQL, // save for resume
      question,
    },
    finalResponse: {
      type: "clarification",
      message: question,
    },
  };
}
